Feat/cordon, drain

pranludi opened the pull request:
- NodeMaintenance crd
- node cordon, uncordon
- pod drain

- test node.yaml
```yaml
apiVersion: project.aipub.ten1010.io/v1alpha1
kind: NodeMaintenance
metadata:
  name: node1-maintenance
spec:
  targetNodes:
    - minikube-m02
#    - minikube-m03
  actions:
#    - type: uncordon
    - type: drain
      ignoreDaemonSets: false
      force: true

```

Commit:
chore

Commit:
wip

Commit:
wip

Commit:
kubectl 으로 특정 노드를 cordon 하면 해당 node 의 pod 을 모두 삭제

Commit:
wip

Commit:
wip

Commit:
rollback - files

Commit:
wip - daemonset 에 대한 대응

daemonset 테스트 팟 등록 pod3.yaml

Commit:
wip - uncordon 지금은 watch 가 작동될때 체크해서 진행하도록 되어 있음
NodeMaintenance, NodeMaintenanceClaim 이 작동되면 체크하도록 변경 필요

Commit:
maintenance yaml 으로 cordon, drain

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix
- targetNode -> targetNodes List type
- add force option

Commit:
fix
- node maintenance status

Commit:
fix
- status

juu-n-oo commented on src/main/java/io/ten1010/aipub/projectcontroller/controller/cr/NodeMaintenanceReconciler.java:
```diff
@@ -0,0 +1,180 @@
+    private final Indexer<V1alpha1NodeMaintenance> projectIndexer;
+    private final CoreV1Api coreV1Api;
+    private final StatusPatchHelper<V1alpha1NodeMaintenance> statusPatchHelper;
+    private final String namespace = "project-controller";
```
임시로 namespaced 범위로 하는것?

juu-n-oo commented on src/main/java/io/ten1010/aipub/projectcontroller/controller/cr/NodeMaintenanceReconciler.java:
```diff
@@ -0,0 +1,180 @@
+
+        List<String> effectedNodes = new ArrayList<>();
+        for (String targetNodeName : targetNodeNames) {
+            var targetNodeRequest = coreV1Api.readNode(targetNodeName);
```
node indexer

juu-n-oo commented on src/main/java/io/ten1010/aipub/projectcontroller/controller/cr/NodeMaintenanceReconciler.java:
```diff
@@ -0,0 +1,180 @@
+                        return new Result(false);
+                    }
+
+                    var podRequest = coreV1Api.listNamespacedPod(namespace);
```
해당 리소스에 대한 reconciler에서 수행

juu-n-oo commented on kubernetes/controller/project-controller/templates/crd.yaml:
```diff
@@ -189,6 +189,75 @@ spec:
+                      default: false
+                status:
+                  type: string
+---
```
1개의 NodeMaintence에 여러 action을 표현할지, 
action마다 NodeMaintence를 생성할지

Commit:
fix
- NodeMaintenance 가 작동시 node, pod 에서 처리하도록 변경

Commit:
fix
- System.out.print 제거

Commit:
fix
- `//` 정리

Commit:
fix

Commit:
fix

Commit:
fix
BoundObjectResolver 에 있던 로직은 NodeReconciler, PodReconciler 으로 이동

Commit:
fix

Commit:
fix
- replicaSet 테스트 yaml 추가
- replicaSet 상태에서 drain 시 종료 대기 시간을 위해 Duration.ofSeconds(1) 으로 변경

Commit:
fix

Commit:
fix

Commit:
fix
- Constants

Commit:
fix
- Constants

Commit:
fix

Commit:
admission controller
